### Segmentation & Clustering_Neighborhood - Week 3 Assignment

In [14]:
import numpy as np 
import json 
from geopy.geocoders import Nominatim
import requests as req
from pandas.io.json import json_normalize 
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
from bs4 import BeautifulSoup as bs
import xml
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
!conda install -c conda-forge folium=0.5.0 --yes 
import folium 

Solving environment: done

# All requested packages already installed.



### Getting Data from Wikipedia

In [15]:
url = req.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = bs(url,'lxml')

### Cleaning Data

In [17]:
table_post = soup.find('table')
fields = table_post.find_all('td')

postcode = []
borough = []
neighbourhood = []

for i in range(0, len(fields), 3):
    postcode.append(fields[i].text.strip())
    borough.append(fields[i+1].text.strip())
    neighbourhood.append(fields[i+2].text.strip())
        
df = pd.DataFrame(data=[postcode, borough, neighbourhood]).transpose()
df.columns = ['Postcode', 'Borough', 'Neighbourhood']
df['Borough'].replace('Not assigned', np.nan, inplace=True)
df.dropna(subset=['Borough'], inplace=True)
df_new = df.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(', '.join).reset_index()
df_new.columns = ['Postcode', 'Borough', 'Neighbourhood']
df_new['Neighbourhood'].replace('Not assigned', "Queen's Park", inplace=True)
df_new

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood  
0                                       Rouge, Malvern  
1               Highland Creek, Rouge Hill, Port Union  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6          East Birchmount Park, Ionview, Kennedy Park  
7          

In [4]:
df_new.shape

(103, 3)

### Finding the latitiude - longitude values of TORONTO CANADA

In [8]:
df_geo = pd.read_csv('http://cocl.us/Geospatial_data')
df_geo.columns = ['Postcode', 'Latitude', 'Longitude']

In [19]:
df2 = pd.merge(df_new, df_geo, on=['Postcode'], how='inner')

df_new = df2[['Borough', 'Neighbourhood', 'Postcode', 'Latitude', 'Longitude']].copy()

df_new

Borough                                      Neighbourhood  \
0         Scarborough                                     Rouge, Malvern   
1         Scarborough             Highland Creek, Rouge Hill, Port Union   
2         Scarborough                  Guildwood, Morningside, West Hill   
3         Scarborough                                             Woburn   
4         Scarborough                                          Cedarbrae   
5         Scarborough                                Scarborough Village   
6         Scarborough        East Birchmount Park, Ionview, Kennedy Park   
7         Scarborough                    Clairlea, Golden Mile, Oakridge   
8         Scarborough    Cliffcrest, Cliffside, Scarborough Village West   
9         Scarborough                        Birch Cliff, Cliffside West   
10        Scarborough  Dorset Park, Scarborough Town Centre, Wexford ...   
11        Scarborough                                  Maryvale, Wexford   
12        Scarborough                                          Agincourt   
13        Scarborough            Clarks Corners, Sullivan, Tam O'Shanter   
14        Scarborough  Agincourt North, L'Amoreaux East, Milliken, St...   
15        Scarborough                                    L'Amoreaux West   
16        Scarborough                                        Upper Rouge   
17         North York                                  Hillcrest Village   
18         North York                       Fairview, Henry Farm, Oriole   
19         North York                                    Bayview Village   
20         North York                           Silver Hills, York Mills   
21         North York                            Newtonbrook, Willowdale   
22         North York                                   Willowdale South   
23         North York                                    York Mills West   
24         North York                                    Willowdale West   
25         North York                                          Parkwoods   
26         North York                                    Don Mills North   
27         North York                   Flemingdon Park, Don Mills South   
28         North York    Bathurst Manor, Downsview North, Wilson Heights   
29         North York                    Northwood Park, York University   
30         North York                        CFB Toronto, Downsview East   
31         North York                                     Downsview West   
32         North York                                  Downsview Central   
33         North York                                Downsview Northwest   
34         North York                                   Victoria Village   
35          East York                    Woodbine Gardens, Parkview Hill   
36          East York                                   Woodbine Heights   
37       East Toronto                                        The Beaches   
38          East York                                            Leaside   
39          East York                                   Thorncliffe Park   
40          East York                                       East Toronto   
41       East Toronto                       The Danforth West, Riverdale   
42       East Toronto                     The Beaches West, India Bazaar   
43       East Toronto                                    Studio District   
44    Central Toronto                                      Lawrence Park   
45    Central Toronto                                   Davisville North   
46    Central Toronto                                 North Toronto West   
47    Central Toronto                                         Davisville   
48    Central Toronto                        Moore Park, Summerhill East   
49    Central Toronto  Deer Park, Forest Hill SE, Rathnelly, South Hi...   
50   Downtown Toronto                                           Rosedale   
51   Downtown Toronto                        Cabbagetown, St. 

### Drawing the map of Toronto with 103 higlited locations

In [20]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_new['Latitude'], df_new['Longitude'], df_new['Borough'], df_new['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=3,
        popup=label,
        color='yellow',
        fill=True,
        fill_color='#eefff',
        fill_opacity=0.3,
        parse_html=False).add_to(map_toronto)  
    
map_toronto